casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala

```diff
@@ -0,0 +1,477 @@
+                  protocol
+                })
+        standalone <- Ref.of[F, CasperPacketHandlerInternal[F]](new StandaloneCasperHandler[F](abp))
+        _ <- {
+          val _ = toTask(
+            StandaloneCasperHandler
+              .approveBlockInterval(conf.approveGenesisInterval,
+                                    runtimeManager,
+                                    validatorId,
+                                    standalone)).forkAndForget.runAsync
+          ().pure[F]
+        }
+      } yield new CasperPacketHandlerImpl[F](standalone)
+    } else {
+      for {
+        validators  <- CasperConf.parseValidatorsFile[F](conf.knownValidatorsFile)
+        validatorId <- ValidatorIdentity.fromConfig[F](conf)
+        bootstrap <- Ref.of[F, CasperPacketHandlerInternal[F]](
+                      new BootstrapCasperHandler(runtimeManager, validatorId, validators))
+        casperPacketHandler = new CasperPacketHandlerImpl[F](bootstrap)
+        _ <- {
+          implicit val ph: PacketHandler[F] = PacketHandler.pf[F](casperPacketHandler.handle)
+          val rb                            = CommUtil.requestApprovedBlock[F](delay)
+          toTask(rb).forkAndForget.runAsync
+          ().pure[F]
+        }
+      } yield casperPacketHandler
+    }
+
+  trait CasperPacketHandlerInternal[F[_]] {
+    def handleBlockMessage(bm: BlockMessage): F[Option[Packet]]
+
+    def handleBlockRequest(peer: PeerNode, br: BlockRequest): F[Option[Packet]]
+
+    def handleApprovedBlock(ab: ApprovedBlock): F[(Option[MultiParentCasper[F]], Option[Packet])]
+
+    def handleApprovedBlockRequest(peer: PeerNode, br: ApprovedBlockRequest): F[Option[Packet]]
+
+    def handleUnapprovedBlock(peer: PeerNode, ub: UnapprovedBlock): F[Option[Packet]]
+
+    def handleBlockApproval(ba: BlockApproval): F[Option[Packet]]
+
+    def handleNoApprovedBlockAvailable(na: NoApprovedBlockAvailable): F[Option[Packet]]
+  }
+
+  /** Node in this state is a genesis block validator. It will respond only to
+    * [[UnapprovedBlock]] messages forwarding the logic of handling this message to
+    * instance of [[BlockApproverProtocol]] class.
+    *
+    * When in this state node can't handle any other message type so it will return `F[None]`
+    **/
+  private[comm] class GenesisValidatorHandler[
+      F[_]: Capture: Monad: NodeDiscovery: TransportLayer: Log: Time: ErrorHandler: RPConfAsk](
+      blockApprover: BlockApproverProtocol)
+      extends CasperPacketHandlerInternal[F] {
+    private val nonePacket: F[Option[Packet]] = Monad[F].pure(None: Option[Packet])
+
+    override def handleBlockMessage(bm: BlockMessage): F[Option[Packet]] = nonePacket
+    override def handleBlockRequest(peer: PeerNode, br: BlockRequest): F[Option[Packet]] =
+      nonePacket
+    override def handleApprovedBlock(
+        ab: ApprovedBlock): F[(Option[MultiParentCasper[F]], Option[Packet])] =
+      nonePacket.map(none[MultiParentCasper[F]] -> _)
```

Comment:
Genesis validators should transition to the normal packet handler state after getting an approved block.

Reply:
Yes, but I think that you said to do it in separate PR and that node won't end up as genesis validator _by mistake_.